test/performance/odp_ipsecfwd.c

```diff
@@ -0,0 +1,1307 @@
+
+	while (true) {
+		opt = getopt_long(argc, argv, shortopts, longopts, &long_index);
+
+		if (opt == -1)
+			break;
+
+		switch (opt) {
+		case 'i':
+			parse_interfaces(config, optarg);
+			break;
+		case 'n':
+			config->num_pkts = atoi(optarg);
+			break;
+		case 'l':
+			config->pkt_len = atoi(optarg);
+			break;
+		case 'c':
+			config->num_thrs = atoi(optarg);
+			break;
+		case 'm':
+			config->mode = !!atoi(optarg);
+			break;
+		case 's':
+			config->sa_conf_file = strdup(optarg);
+			break;
+		case 'f':
+			config->fwd_conf_file = strdup(optarg);
+			break;
+		case 'h':
+			print_usage();
+			return PRS_TERM;
+		case '?':
+		default:
+			print_usage();
+			return PRS_NOK;
+		}
+	}
+
+	parse_sas(config);
+	parse_fwd_table(config);
+	free(config->sa_conf_file);
+	free(config->fwd_conf_file);
+
+	return check_options(config);
+}
+
+static parse_result_t setup_program(int argc, char **argv, prog_config_t *config)
+{
+	struct sigaction action = { .sa_handler = terminate };
+
+	if (sigemptyset(&action.sa_mask) == -1 || sigaddset(&action.sa_mask, SIGINT) == -1 ||
+	    sigaddset(&action.sa_mask, SIGTERM) == -1 ||
+	    sigaddset(&action.sa_mask, SIGHUP) == -1 || sigaction(SIGINT, &action, NULL) == -1 ||
+	    sigaction(SIGTERM, &action, NULL) == -1 || sigaction(SIGHUP, &action, NULL) == -1) {
+		ODPH_ERR("Error installing signal handler\n");
+		return PRS_NOK;
+	}
+
+	init_config(config);
+
+	return parse_options(argc, argv, config);
+}
+
+static int send(const pktio_t *pktio, uint8_t index, odp_packet_t pkt)
+{
+	return odp_pktout_send(pktio->out_dir_qs[index], &pkt, 1);
+}
+
+static int enqueue(const pktio_t *pktio, uint8_t index ODP_UNUSED, odp_packet_t pkt)
+{
+	return odp_queue_enq(pktio->out_ev_q, odp_packet_to_event(pkt));
+}
+
+static odp_bool_t setup_pktios(prog_config_t *config)
+{
+	odp_pool_param_t pool_param;
+	pktio_t *pktio;
+	odp_pktio_param_t pktio_param;
+	odp_pktin_queue_param_t pktin_param;
+	odp_pktio_capability_t capa;
+	odp_pktout_queue_param_t pktout_param;
+	odp_pktio_config_t pktio_config;
+	uint32_t num_tx_qs;
+
+	odp_pool_param_init(&pool_param);
+	pool_param.pkt.seg_len = config->pkt_len;
+	pool_param.pkt.len = config->pkt_len;
+	pool_param.pkt.num = config->num_pkts;
+	pool_param.type = ODP_POOL_PACKET;
+	config->pktio_pool = odp_pool_create(PROG_NAME, &pool_param);
+
+	if (config->pktio_pool == ODP_POOL_INVALID) {
+		ODPH_ERR("Error creating packet I/O pool\n");
+		return false;
+	}
+
+	for (uint32_t i = 0U; i < config->num_ifs; ++i) {
+		pktio = &config->pktios[i];
+		odp_pktio_param_init(&pktio_param);
+		pktio_param.in_mode = ODP_PKTIN_MODE_SCHED;
+		pktio_param.out_mode = config->mode == 0U ? ODP_PKTOUT_MODE_QUEUE :
+							    ODP_PKTOUT_MODE_DIRECT;
+		pktio->handle = odp_pktio_open(pktio->name, config->pktio_pool, &pktio_param);
+
+		if (pktio->handle == ODP_PKTIO_INVALID) {
+			ODPH_ERR("Error opening packet I/O (%s)\n", pktio->name);
+			return false;
+		}
+
+		odp_pktin_queue_param_init(&pktin_param);
```

Comment:
This sets pktin_param.queue_param.sched.sync to the default value ODP_SCHED_SYNC_PARALLEL, but I think we should use ODP_SCHED_SYNC_ORDERED.

Reply:
Changed in v4 and now depends on the configured `mode`.